File: backend/schemas/workbench.py
"""
Workbench Core Schemas

Core business object schemas for the workbench functionality
These must align exactly with the frontend types in workbench.ts
"""

from typing import List, Optional, Dict, Any, Literal
from pydantic import BaseModel, Field

from schemas.canonical_types import CanonicalResearchArticle


# ================== FEATURE METADATA AND DATA STRUCTURES ==================

class FeatureDefinition(BaseModel):
    """Feature metadata stored in article groups - defines what features to extract"""
    id: str = Field(..., description="Stable UUID for feature identification")
    name: str = Field(..., description="Feature display name") 
    description: str = Field(..., description="Feature description for LLM extraction")
    type: Literal['boolean', 'text', 'score'] = Field(..., description="Feature data type")
    options: Optional[Dict[str, Any]] = Field(None, description="Feature options (e.g., min/max for score)")



# Note: Feature data is stored directly in ArticleGroupDetail.feature_data
# No separate column data structure needed

# ================== ARTICLE GROUP STRUCTURES ==================

class ArticleGroupDetail(BaseModel):
    """Individual article item within a group context - junction model"""
    id: str = Field(..., description="Unique detail record ID")
    article_id: str = Field(..., description="Article ID")
    group_id: str = Field(..., description="Group ID")
    article: CanonicalResearchArticle = Field(..., description="The article data")
    feature_data: Dict[str, Any] = Field(default_factory=dict, description="Extracted feature data keyed by feature.id")
    position: Optional[int] = Field(None, description="Position in the group")
    added_at: str = Field(..., description="When article was added to group")


class PaginationInfo(BaseModel):
    """Pagination metadata"""
    current_page: int = Field(..., description="Current page number")
    total_pages: int = Field(..., description="Total number of pages")
    total_results: int = Field(..., description="Total number of results")
    page_size: int = Field(..., description="Number of items per page")

class ArticleGroupWithDetails(BaseModel):
    """Complete article group with embedded articles - full group context"""
    id: str = Field(..., description="Group ID")
    name: str = Field(..., description="Group name")
    description: Optional[str] = Field(None, description="Group description")
    article_count: int = Field(..., description="Number of articles in group")
    feature_definitions: List[FeatureDefinition] = Field(..., description="Feature definitions")
    search_context: Optional[Dict[str, Any]] = Field(None, description="Search context")
    created_at: str = Field(..., description="Creation timestamp")
    updated_at: str = Field(..., description="Last update timestamp")
    articles: List[ArticleGroupDetail] = Field(..., description="Articles with feature data")
    pagination: Optional[PaginationInfo] = Field(None, description="Pagination metadata")


class ArticleGroup(BaseModel):
    """Base article group info - lightweight group metadata"""
    id: str = Field(..., description="Group ID")
    user_id: int = Field(..., description="Owner user ID")
    name: str = Field(..., description="Group name")
    description: Optional[str] = Field(None, description="Group description")
    search_query: Optional[str] = Field(None, description="Search query used")
    search_provider: Optional[str] = Field(None, description="Search provider used")
    search_params: Optional[Dict[str, Any]] = Field(None, description="Search parameters")
    feature_definitions: List[FeatureDefinition] = Field(default_factory=list, description="Feature definitions")
    article_count: int = Field(..., description="Number of articles in group")
    created_at: str = Field(..., description="Creation timestamp")
    updated_at: str = Field(..., description="Last update timestamp")


# ================== ADDITIONAL BACKEND TYPES ==================
# Note: These types exist only in the frontend:
# - WorkbenchState (frontend UI state)
# - ArticleGroupSummary (frontend display type)
# - ArticlePreview (frontend display type)
# - WorkbenchData (individual article research)
# - ExtractedFeature (individual article features)
# - WorkbenchMetadata (individual article metadata)
# - AnalysisPreset (analysis presets)

File: frontend/src/types/workbench.ts
/**
 * Workbench Core Types
 * 
 * Core business object types for the workbench functionality
 * These must align exactly with the backend schemas in workbench.py
 */

import { CanonicalResearchArticle } from './canonical_types';

// ================== COLUMN METADATA AND DATA STRUCTURES ==================

export interface WorkbenchColumnMetadata {
  name: string;
  description: string;
  type: 'boolean' | 'text' | 'score' | 'number';
  options?: {
    min?: number;
    max?: number;
    step?: number;
    choices?: string[];
  };
  is_extracted: boolean;
  extraction_method?: 'ai' | 'manual' | 'computed';
}

export interface WorkbenchColumn {
  id: string;
  name: string;
  description: string;
  type: 'boolean' | 'text' | 'score' | 'number';
  data: Record<string, string>; // articleId -> value
  options?: {
    min?: number;
    max?: number;
    step?: number;
    choices?: string[];
  };
}

// Note: TabelizerColumnData is internal to backend - frontend uses WorkbenchColumn

// ================== ARTICLE GROUP STRUCTURES ==================

export interface ArticleGroupItem {
  article: CanonicalResearchArticle;
  position: number;
  column_data: Record<string, any>;
  workbench_summary: {
    has_notes: boolean;
    feature_count: number;
    tags: string[];
    rating?: number;
  };
}

export interface ArticleGroupDetail {
  id: string;
  name: string;
  description?: string;
  article_count: number;
  columns: WorkbenchColumnMetadata[];
  search_context?: {
    query: string;
    provider: string;
    parameters: Record<string, any>;
  };
  created_at: string;
  updated_at: string;
  articles: ArticleGroupItem[];
}

export interface ArticleGroup {
  id: string;
  user_id: number;
  name: string;
  description?: string;
  search_query?: string;
  search_provider?: string;
  search_params?: any;
  columns: WorkbenchColumnMetadata[];
  article_count: number;
  created_at: string;
  updated_at: string;
}

// ================== ADDITIONAL FRONTEND TYPES ==================

export interface WorkbenchState {
  articles: CanonicalResearchArticle[];
  columns: WorkbenchColumn[];
  sortBy?: string;
  sortDirection?: 'asc' | 'desc';
}

export interface ArticleGroupSummary {
  id: string;
  name: string;
  description?: string;
  article_count: number;
  created_at: string;
  updated_at: string;
  preview_articles: ArticlePreview[];
}

export interface ArticlePreview {
  id: string;
  title: string;
  journal?: string;
  publication_year?: number;
}

// ================== INDIVIDUAL ARTICLE RESEARCH TYPES ==================

export interface WorkbenchData {
  article: CanonicalResearchArticle;
  workbench: {
    notes: string;
    features: Record<string, ExtractedFeature>;
    metadata: WorkbenchMetadata;
    position: number;
    created_at: string;
    updated_at?: string;
  };
  group_context: {
    group_id: string;
    group_name: string;
    total_articles: number;
  };
}

export interface ExtractedFeature {
  value: string | number | boolean;
  type: 'boolean' | 'text' | 'score' | 'number';
  extraction_method: 'ai' | 'manual';
  extraction_prompt?: string;
  confidence_score?: number;
  extracted_at: string;
  extracted_by?: string;
  error?: string;
}

export interface WorkbenchMetadata {
  tags: string[];
  rating?: number; // 1-5 stars
  priority: 'low' | 'medium' | 'high';
  status: 'unread' | 'reading' | 'read' | 'reviewed' | 'archived';
  custom_fields?: Record<string, any>;
}

// ================== ANALYSIS PRESET TYPES ==================

export interface AnalysisPreset {
  id: string;
  name: string;
  description: string;
  category: string;
  columns: Record<string, {
    description: string;
    type: 'boolean' | 'text' | 'score' | 'number';
    options?: {
      min?: number;
      max?: number;
      step?: number;
      choices?: string[];
    };
  }>;
  created_at: string;
  usage_count: number;
}

// ================== LEGACY TYPE ALIASES ==================
// For backward compatibility during migration

/** @deprecated Use WorkbenchColumn instead */
export type TabelizerColumn = WorkbenchColumn;

/** @deprecated Use WorkbenchColumnMetadata instead */  
export type TabelizerColumnData = WorkbenchColumnMetadata;

/** @deprecated Use WorkbenchState instead */
export type TabelizerState = WorkbenchState;

/** @deprecated Use AnalysisPreset instead */
export type TabelizerPreset = AnalysisPreset;